### PR TITLE
Handle testhost close

### PR DESCRIPTION
**Pull request by mayankbansal018**

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs [hunk 1]
```diff
@@ -67,7 +67,6 @@ protected ParallelOperationManager(Func<T> createNewManager, int parallelLevel,
         public void RemoveManager(T manager)
         {
             this.concurrentManagerHandlerMap.Remove(manager);
-            this.DisposeInstance(manager);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs]: Remove the implementation and all consumers.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 2]
```diff
@@ -139,13 +141,42 @@ public void Abort()
...
+            this.baseTestDiscoveryEventsHandler.HandleDiscoveredTests(discoveredTestCases);
+        }
+
+        /// <inheritdoc/>
+        public void HandleRawMessage(string rawMessage)
+        {
+            var message = this.dataSerializer.DeserializeMessage(rawMessage);
+            if(string.Equals(message.MessageType, MessageType.DiscoveryComplete))
+            {
+                this.CleanupManager();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: Call this.Close and make Close a protected?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 3]
```diff
@@ -23,13 +23,14 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client
...
     /// </summary>
-    public class ProxyDiscoveryManager : ProxyOperationManager, IProxyDiscoveryManager
+    public class ProxyDiscoveryManager : ProxyOperationManager, IProxyDiscoveryManager, ITestDiscoveryEventsHandler2
     {
         private readonly ITestRuntimeProvider testHostManager;
         private IDataSerializer dataSerializer;
         private CancellationTokenSource cancellationTokenSource;
         private bool isCommunicationEstablished;
         private IRequestData requestData;
+        private ITestDiscoveryEventsHandler2 baseTestDiscoveryEventsHandler;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: Try LUT scenarios

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 4]
```diff
@@ -139,13 +141,42 @@ public void Abort()
...
+        public void HandleDiscoveredTests(IEnumerable<TestCase> discoveredTestCases)
+        {
+            this.baseTestDiscoveryEventsHandler.HandleDiscoveredTests(discoveredTestCases);
+        }
+
+        /// <inheritdoc/>
+        public void HandleRawMessage(string rawMessage)
+        {
+            var message = this.dataSerializer.DeserializeMessage(rawMessage);
+            if(string.Equals(message.MessageType, MessageType.DiscoveryComplete))
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: Invariant check

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyDiscoveryManagerTests.cs [hunk 5]
```diff
@@ -323,5 +328,53 @@ public void DiscoverTestsShouldCollectMetrics()
                 TestPluginCache.Instance = null;
             }
         }
+
+        [TestMethod]
+        public void DiscoverTestsCloseTestHostIfRawMessageIsOfTypeDiscoveryComplete()
+        {
+            Mock<ITestDiscoveryEventsHandler2> mockTestDiscoveryEventsHandler = new Mock<ITestDiscoveryEventsHandler2>();
+            this.mockTestHostManager.Setup(tmh => tmh.LaunchTestHostAsync(It.IsAny<TestProcessStartInfo>(), It.IsAny<CancellationToken>())).Returns(Task.FromResult(true));
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyDiscoveryManagerTests.cs]: Please remove this

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs [hunk 6]
```diff
@@ -189,17 +192,57 @@ public void Abort()
...
+            var message = this.dataSerializer.DeserializeMessage(rawMessage);
+
+            if(string.Equals(message.MessageType, MessageType.ExecutionComplete))
+            {
+                this.CleanupManager();
+            }
+            this.baseTestRunEventsHandler.HandleRawMessage(rawMessage);
+        }
+
+        public void HandleLogMessage(TestMessageLevel level, string message)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs]: Add tests for other public event handlers.

**Review by codito — APPROVED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs (hunk 1 above)
Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs]: resolved

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs (hunk 3 above)
Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: Working

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs (hunk 4 above)
Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: Not Needed

**Pull request merged**